### PR TITLE
Make sure the rounding mode is restored when doing exact computations (5.2)

### PR DESCRIPTION
**Pull request by sloriot**
Thoses changes should have been done when the code was modified for
CGAL_CFG_FPU_ROUNDING_MODE_UNWINDING_VC_BUG

### CODE REVIEW

**Comment by mglisse** [comment]
Should we use Checked_protect_FPU_rounding for the TO_NEAREST before exact computations? (I didn't think about it much)

**Comment by sloriot** [comment]
It is actually a good idea as we recently notice it is important with `boost::cpp_int` and `boost::cpp_rattional`. I'll add that.

File: Filtered_kernel/include/CGAL/Lazy.h [hunk 1]
```diff
@@ -1594,7 +1602,7 @@ struct Lazy_construction_variant {
     }
     CGAL_BRANCH_PROFILER_BRANCH(tmp);
     Protect_FPU_rounding<!Protection> P2(CGAL_FE_TONEAREST);
-
+    CGAL_expensive_assertion(FPU_get_cw() == CGAL_FE_TONEAREST);
```
Comment (mglisse) [inline Filtered_kernel/include/CGAL/Lazy.h]: Why not use Checked_protect_FPU_rounding (as in Interval_nt) which is supposed to do the same as those 2 lines? (although its last assertion seems to be bogus)

**Comment by lrineau** [comment]
@sloriot What is the bug, and when was it introduced?

**Comment by sloriot** [comment]
Years ago, when we discovered the non-restoration of the rounding mode by MSVC when going out of a try block to the catch block. The workaround has been correctly implemented for Filtered_predicates for example but it was not correct at some places. The PRs implements the workaround consistently. We can discuss IRL if you need more info

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.4-Ic-53

**Pull request merged**